### ISSUE TITLE
error: internal compiler error: src/librustc_trans/type_of.rs:154: Unexpected tail in unsized_info_ty: Foo for ty=Foo

### ISSUE THREAD

**Issue by rnewman** [rust-lang/rust#38838]
This small amount of malformed Rust:

```rust
enum Element {}

enum Foo {
    FooArray { elements: [Element] },
    FooElem { element: Element },
}

fn bar(spec: &Foo) -> bool {
    match spec {
        &Foo::FooArray { elements: _ }   => false,
        &Foo::FooElem { element: _ }   => false,
    }
}

pub fn main() {
    print!("Main");
}
```

causes `rustc` 1.13 to panic:

```
error: internal compiler error: src/librustc_trans/type_of.rs:154: Unexpected tail in unsized_info_ty: Foo for ty=Foo
…
thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:656
stack backtrace:
   1:        0x106579488 - std::sys::backtrace::tracing::imp::write::h6f1d53a70916b90d
   2:        0x106586c4f - std::panicking::default_hook::{{closure}}::h137e876f7d3b5850
   3:        0x1065840fd - std::panicking::default_hook::h0ac3811ec7cee78c
   4:        0x106584796 - std::panicking::rust_panic_with_hook::hc303199e04562edf
   5:        0x1064584da - std::panicking::begin_panic::hf3c77cb5fd5814c4
   6:        0x10646897a - rustc_errors::Handler::bug::h9bb46cf778492ea1
   7:        0x103f5aadc - rustc::session::opt_span_bug_fmt::{{closure}}::hf0e25c590c479ec5
   8:        0x103e90729 - rustc::session::opt_span_bug_fmt::h20b1b2ceeefb3522
   9:        0x103e904da - rustc::session::bug_fmt::h203210307841d95b
  10:        0x1030fb8ae - rustc_trans::type_of::unsized_info_ty::h680ddbd47367bc7c
  11:        0x1030fbf8b - rustc_trans::type_of::in_memory_type_of::h570d089fc2dbacc6
  12:        0x103109522 - rustc_trans::abi::FnType::unadjusted::{{closure}}::h920834572dd7a84e
  13:        0x103061d1f - rustc_trans::abi::FnType::unadjusted::he2b6c17199499f5c
  14:        0x1030b9171 - rustc_trans::declare::declare_fn::h721816f3ca217145
  15:        0x1030f4e9b - rustc_trans::trans_item::TransItem::predefine::hffbc6c5dfea77a18
  16:        0x103073131 - rustc_trans::base::trans_crate::h7b6c2eb43ae0c1da
  17:        0x102ef0d21 - rustc_driver::driver::phase_4_translate_to_llvm::h0646ffc4a1716d29
  18:        0x102f2a3bb - rustc_driver::driver::compile_input::{{closure}}::h5a81b1e8a8f3f531
  19:        0x102f168b2 - rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::h4091eeab6f271a29
  20:        0x102eee9c6 - rustc_driver::driver::phase_3_run_analysis_passes::hff1c31d493acfd09
  21:        0x102edc7da - rustc_driver::driver::compile_input::hef221f9f618e92f5
  22:        0x102f048b8 - rustc_driver::run_compiler::h7dddd4089f8330b4
  23:        0x102e463b0 - std::panicking::try::do_call::h1e3ece12e3f812df
  24:        0x10658720a - __rust_maybe_catch_panic
  25:        0x102e65074 - <F as alloc::boxed::FnBox<A>>::call_box::h05b222e46803eb30
  26:        0x1065833b4 - std::sys::thread::Thread::new::thread_start::h759e10bc4abc7e72
  27:     0x7fff9591399c - _pthread_body
  28:     0x7fff95913919 - _pthread_start
```

**Comment by rnewman** [comment]
Full version details:

```
rustc 1.13.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-apple-darwin
release: 1.13.0
```

**Comment by apasel422** [comment]
This was fixed in 1.14.

**Comment by rnewman** [comment]
Ah, HomeBrew doesn't have 1.14 yet. That explains it!
